src/databricks/labs/dqx/llm/llm_core.py

```diff
@@ -0,0 +1,315 @@
+        model_type="chat",
+        api_key=api_key,
+        api_base=api_base,
+        max_retries=3,
+    )
+    dspy.configure(lm=language_model)
+
+
+def validate_generated_rules(actual: str) -> float:
+    """
+    Validate generated rules with granular scoring for better optimizer feedback.
+
+    Scoring breakdown:
+        - JSON parsing (40%): Checks if the actual output can be parsed as valid JSON.
+        - Rules validation (60%): Ensures the rules pass DQX validation checks.
+
+    Args:
+        actual (str): JSON string of the actual generated rules.
+
+    Returns:
+        float: A score between 0.0 and 1.0 representing the quality of the generated rules.
+    """
+    total_score = 0.0
+
+    # Score weights
+    json_weight = 0.4
+    rule_weight = 0.6
+
+    # Json parsing score (40%)
+    try:
+        actual_rules = json.loads(actual)
+        total_score += json_weight
+        logger.debug(f"✓ JSON parsing successful (+{json_weight:.1f})")
+    except json.JSONDecodeError as e:
+        logger.warning(f"✗ JSON parsing failed: {e}")
+        logger.debug(f"  Raw output: {repr(actual[:200])}")
+        # Early return if we can't parse JSON at all
+        return total_score
+
+    # Rules validation score (60%)
+    validation_status = DQEngineCore.validate_checks(actual_rules)
+    if not validation_status.has_errors:
+        total_score += rule_weight
+        logger.debug(f"✓ Rules validation passed (+{rule_weight:.1f})")
+    else:
+        logger.warning(f"✗ Rules validation errors: {validation_status.errors}")
+
+    logger.debug(f"Final score: {total_score:.2f}")
+    return total_score
+
+
+def get_dspy_compiler(
+    model: str = "databricks/databricks-meta-llama-3-3-70b-instruct",
+    api_key: str = "",
+    api_base: str = "",
+    enable_schema_inference: bool = True,
+) -> dspy.Module:
+    """
+    Get the Dspy compiler configured with an optimizer.
+
+    This function initializes and configures the Dspy compiler with a training set and an optimizer
+    to validate and optimize the generated data quality rules.
+
+    Args:
+        model (str): The model to use for the Dspy language model.
+        api_key (str): The API key for the model. Not required by Databricks foundational models.
+        api_base (str): The API base URL for the model. Not required by Databricks foundational models.
+        enable_schema_inference (bool): If True, the model will infer schema when not provided.
+
+    Returns:
+        dspy.Module: An optimized Dspy module for generating data quality rules.
+    """
+    _configure_dspy_model(api_key=api_key, api_base=api_base, model=model)
+
+    # Use standard DSPy approach with improved prompting
+    dq_model = DQRuleGeneration(enable_schema_inference=enable_schema_inference)
+    train_set = create_optimizer_training_set()
+
+    # Standard metric for JSON output validation
+    def json_metric(_example, pred, _trace=None):
+        if hasattr(pred, 'quality_rules'):
```

Comment:
```suggestion
        if pred.quality_rules:
```
don't use reflection if you can access the field directly